migration/20220428-namespace-overdrive-configuration.py

```diff
@@ -0,0 +1,174 @@
+    if result_new_key.rowcount > 0:
+        existing_new_key_present = True
+        existing_new_key = result_new_key.fetchone()[0]
+
+    if existing_old_key_present:
+        logger.info(f"Discovered old key '{old_key}' -> '{existing_old_key}'")
+    if existing_new_key_present:
+        logger.info(f"Discovered new key '{new_key}' -> '{existing_new_key}'")
+
+    # If keys have a non-null value, something is seriously wrong!
+    if not existing_old_key and not existing_new_key:
+        raise RuntimeError(
+            f"External integration {integration} has neither a '{old_key}' or a '{new_key}'"
+        )
+
+    # If both the old key and the new key exist, then keep whichever one is non-null,
+    # preferring the new key to the old key.
+    if existing_old_key_present and existing_new_key_present:
+        if existing_new_key:
+            logger.info(f"Deleting old key '{old_key}'")
+            delete_key(db=db, integration=integration, key=old_key)
+        else:
+            assert existing_old_key
+            rename_key(db=db, integration=integration, old_key=old_key, new_key=new_key)
+        return 1
+
+    # Otherwise, if only the old key exists, rename it to the new key.
+    if existing_old_key_present:
+        assert not existing_new_key
+        assert not existing_new_key_present
+        assert existing_old_key
+
+        logger.info(f"Renaming old key '{old_key}' -> new key '{new_key}'")
+        rename_key(db=db, integration=integration, old_key=old_key, new_key=new_key)
+        return 1
+
+    # The old key didn't exist, the new key exists, so there's nothing to do.
+    assert existing_new_key
+    assert existing_new_key_present
+    assert not existing_old_key
+    assert not existing_old_key_present
+
+    logger.info(
+        f"New key '{new_key}' already exists, and old key '{old_key}' does not; nothing to do!"
+    )
+    return 0
+
+
+def update_configuration(db: Session, integration: int) -> None:
+    updated = 0
+    updated += update_key_value(
+        db=db,
+        integration=integration,
+        old_key="username",
+        new_key="overdrive_client_key",
+    )
+    updated += update_key_value(
+        db=db,
+        integration=integration,
+        old_key="password",
+        new_key="overdrive_client_secret",
+    )
+    updated += update_key_value(
+        db=db,
+        integration=integration,
+        old_key="website_id",
+        new_key="overdrive_website_id",
+    )
+    updated += update_key_value(
+        db=db,
+        integration=integration,
+        old_key="server_nickname",
+        new_key="overdrive_server_nickname",
+    )
+
+    if updated > 0:
+        logger.info("Updates were made, committing...")
+        db.commit()
+    else:
+        logger.info("No updates were necessary")
+
+
+def execute_migration(db: Session) -> None:
+    select = text(
+        """
+SELECT e.id FROM externalintegrations AS e
+  WHERE e.protocol = 'Overdrive'
+    """
```

Comment:
The properties enumerated in `update_configuration` are not associated with child OverDrive accounts (i.e., those that have a parent collection, like OverDrive Advantage collections). So we should not try to update the keys for OD collections that have a parent collection. Changing this query to the following should get the correct set of the `externalintegrations`:
```
    select = text(
        """
        SELECT e.id
        FROM collections c
        JOIN externalintegrations e on c.external_integration_id = e.id
        WHERE e.protocol = 'Overdrive' AND c.parent_id IS NULL
        """
    )
```

Reply:
You're not wrong, but I'm not sure this is necessary. The purpose of this migration is solely to stop using the old names (Use `overdrive_client_key` instead of `username` and so on). Unless I'm mistaken, it shouldn't matter whether a collection has a parent or not when it comes to which names it's using.

Reply:
We fail the migration with a `RuntimeError` (ll. 81-85 above), if both old and new keys are missing. Because neither of those keys will be present for an OD Advantage integration, the migration would fail for any CM with any collection of that type.